File: languages/go/02-core/algorithms-basics/example/best-streak.go
// This helper example focuses on isolating one pass over the data so the main example can explain the result clearly.

// This extra example extends algorithms basics with a best-streak scan.

package main

import "fmt"

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
func longestRisingStreak(values []int) (int, []int) {
	if len(values) == 0 {
		return 0, []int{}
	}

	bestLength := 1
	bestStart := 0
	currentLength := 1
	currentStart := 0

	for index := 1; index < len(values); index++ {
		if values[index] > values[index-1] {
			currentLength++
		} else {
			currentLength = 1
			currentStart = index
		}

		if currentLength > bestLength {
			bestLength = currentLength
			bestStart = currentStart
		}
	}

	streak := append([]int{}, values[bestStart:bestStart+bestLength]...)
	return bestLength, streak
}

func main() {
	temperatures := []int{17, 18, 21, 19, 20, 22, 25, 24}
	length, streak := longestRisingStreak(temperatures)

	fmt.Printf("Longest rising streak length: %d\n", length)
	fmt.Printf("Streak values: %v\n", streak)
}

File: languages/csharp/02-core/algorithms-basics/example/best-streak.cs
// This helper example focuses on isolating one pass over the data so the main example can explain the result clearly.

// This extra example extends algorithms basics with a best-streak scan.

using System;
using System.Collections.Generic;

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
class Program
{
    static List<int> LongestRisingStreak(List<int> values)
    {
        if (values.Count == 0)
        {
            return new List<int>();
        }

        int bestLength = 1;
        int bestStart = 0;
        int currentLength = 1;
        int currentStart = 0;

        for (int index = 1; index < values.Count; index++)
        {
            if (values[index] > values[index - 1])
            {
                currentLength++;
            }
            else
            {
                currentLength = 1;
                currentStart = index;
            }

            if (currentLength > bestLength)
            {
                bestLength = currentLength;
                bestStart = currentStart;
            }
        }

        return values.GetRange(bestStart, bestLength);
    }

    static void Main()
    {
        List<int> temperatures = new List<int> { 17, 18, 21, 19, 20, 22, 25, 24 };
        List<int> streak = LongestRisingStreak(temperatures);

        Console.WriteLine($"Longest rising streak length: {streak.Count}");
        Console.WriteLine($"Streak values: [{string.Join(", ", streak)}]");
    }
}
